react components (#36)
Ada-Developers-Academy/core-unit-3: react-components/jsx.md

# JSX

## Goals

React JS can use JavaScript syntax known as JSX. It's used when rendering components in our webapp, so we can expect to see a lot of JSX!

JSX is used to declare HTML-like UI in JavaScript. JSX determines what our webapp's rendered HTML looks like.

> Before we use JSX extensively, let's pause to understand what JSX is and how to write it.

Author: this is awesome-- a really good and powerful framing of JSX. This kind of thing is tricky, so I appreciate the delicacy here!

## Introduction

In a newly created React app using `create-react-app`, we can see the following code snippets:

`src/index.js`

```js
<React.StrictMode>
  <App />
</React.StrictMode>
```

`src/App.js`

```js
<div className="App">
  <header className="App-header">
    <img src={logo} className="App-logo" alt="logo" />
    <p>
      Edit <code>src/App.js</code> and save to reload.
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      target="_blank"
      rel="noopener noreferrer"
    >
      Learn React
    </a>
  </header>
</div>
```

What is this code? How do we read it? How do we write it?

## JSX

> [**JSX**](https://reactjs.org/docs/introducing-jsx.html) is a _syntax extension_ to JavaScript. It is JavaScript code that isn't normally available in vanilla JS projects.
> 
> In our React projects, we'll write the logic for _how_ to render our components in JSX.

Reviewer: i think it reads a bit more easily if we combine the two sentences about the syntax. And I don't know why, but I like the idea of adding this sneaky note about jsx really being js, since it is, from a certain point of view.
```suggestion
[**JSX**](https://reactjs.org/docs/introducing-jsx.html) is a _syntax extension_ to JavaScript that isn't normally available in vanilla JavaScript projects. It may look like HTML, but it's really JavaScript _in disguise_!

In our React projects, we'll write the logic for _how_ to render our components using JSX.
```

## Objects in JSX

We will create _objects_ in JSX.

An object in JSX contains at least one element.

```js

> const welcomeMessage = <span>Welcome!</span>;

Reviewer: for consistency with the description
```suggestion
const welcomeMessage = <span>Welcome~</span>;
```

```

> | <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                            |

Reviewer: widen layout
```suggestion
| <div style="min-width:230px;"> Piece of Code </div> | Notes                                                                                                                            |
```

        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}
```

The `App` function returns one JSX object. This JSX object opens with a `<div>` element.

This `<div>` element has one child: the `<header>` element. The `<header>` element has three children: `<img>`, `<p>`, and `<a>`.

### !callout-info

## Modifying `App` to Use Variables

We can modify our `App` function to use more variables. The JSX object can be defined inside the function, and then returned.

```js
function App() {
  const appElements = <div className="App">... more elements ...</div>;
  return appElements;
}
```

Here, we are creating the `const` variable `appElements`, setting it to a JSX object, and then returning it.

### !end-callout

## Embedding Expressions in JSX

When we have a JavaScript expression, such as a variable or function call, to add into a JSX object, we can do so with `{}`.

> 

Reviewer: ```suggestion

<!-- prettier-ignore-start -->
```

```js

> const guestName = "Megha";

Reviewer: ```suggestion
const guestName = 'Megha';
```

const welcomeMessage = (
  <section>
    <h1>Welcome {guestName}~</h1>
    <p>It's nice to hear from you again!</p>
  </section>
);
```

> 

Reviewer: ```suggestion
<!-- prettier-ignore-end -->

```

In this situation, the rendered HTML would look like:

```html
<section>
  <h1>Welcome Megha~</h1>
  <p>It's nice to hear from you again!</p>
</section>
```

### Embedding Expressions: `App.js`

Let's revisit this `src/App.js` again!

> 

Reviewer: ```suggestion

<!-- prettier-ignore-start -->
```

```js

> import logo from "./logo.svg";
> import "./App.css";

Reviewer: ```suggestion
import logo from './logo.svg';
import './App.css';
```

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}
```

> 

Reviewer: ```suggestion
<!-- prettier-ignore-end -->

```

We can focus on this part of the JSX:

```js
<img src={logo} className="App-logo" alt="logo" />
```

The value of the `src` attribute for this element will be the value of `logo`. `logo` was imported from `./logo.svg` at the top of `App.js`!

## Element Attributes in JSX Objects

> We can define any attributes for any element in a JSX object. We use syntax that is nearly identical to what we'd use for HTML elements.

Reviewer: stay explicitly on the topic of HTML element attributes, otherwise this veers dangerously close to props.
I think we can phrase the syntax issue even more strongly.
```suggestion
We can set the element attributes for JSX objects just like we can for the HTML elements they resemble. For most element attributes, the JSX object syntax is identical to what we'd use for HTML elements.
```

```js
<a href="https://reactjs.org" target="_blank" rel="noopener noreferrer">
  Learn React
</a>
```

This `<a>` element will have the following HTML attributes when it's rendered:

- `href`
- `target`
- `rel`

## CSS Classes in JSX Objects

> One of the most important differences between JSX objects and HTML elements is how to add classes.
> 
> In JSX, we use the attribute `className` to define `class` in rendered HTML.

Reviewer: make sure we're clear about what kind of classes we're discussing
the reader is familiar with HTML class attributes, so lead with that
```suggestion
One of the most important differences between JSX objects and HTML elements is how to add CSS classes.

While HTML elements use the `class` attribute, in JSX we use the `className` attribute.
```

For example, this JSX object:

```js
<header className="App-header"></header>
```

will ultimately be rendered as:

```html
<header class="App-header"></header>
```

This will be crucial for styling!

## Comments in JSX Syntax

> A JSX object can contain code comments only if the comments are embedded as a JS expression.

Reviewer: ```suggestion
A JSX object can contain code comments only if the comments are embedded as a JavaScript expression.
```
